### PR TITLE
#isArg add comment to tell users to not use it

### PR DESCRIPTION
**Pull request by hogoww**
isArg and isTemp have the same convention but they express different things.
there is both isArg that is the Compiler's view on the AST (and therefore relies on compiler stuff)
and isArgument which expresses the AST point of view.

isTemp is (as of this morning) not relying on the compiler.

So naming is still bad, but at least there is a little bit of explanation.

Also removes an  (I'm pretty sure) erroneous user of isArg

### CODE REVIEW

File: src/OpalCompiler-Core/RBVariableNode.extension.st [hunk 1]
```diff
@@ -13,7 +13,8 @@ RBVariableNode >> binding: aSemVar [
 
 { #category : #'*opalcompiler-core' }
 RBVariableNode >> isArg [
-
+	"This is a compiler specific function, requiring a binding.
+	If you're working on the AST, you should use probably isArgument"
```
Comment (guillep) [inline src/OpalCompiler-Core/RBVariableNode.extension.st:17]: I've just checked at a non-semantically analysed AST. I did

```smalltalk
method parseTree.
```
And I got

image

If you look, it does not use argument nodes. So actually `#isArg` and `#isArgument` are both compiler dependent. None of them is valid without a semantic analysis. The first one requires a binding, the second one requires the replacement of the node...

File: src/GeneralRules/ReAssignmentOnBlockArgumentRule.class.st [hunk 2]
```diff
@@ -32,7 +32,7 @@ ReAssignmentOnBlockArgumentRule class >> uniqueIdentifierName [
 ReAssignmentOnBlockArgumentRule >> check: aMethod forCritiquesDo: aCriticBlock [
 	| problemAssignments|
 	problemAssignments := aMethod ast assignmentNodes 
-		select: [ :assignment | assignment variable isArg].
+		select: [ :assignment | assignment variable isArgument ].
```
Comment (guillep) [inline src/GeneralRules/ReAssignmentOnBlockArgumentRule.class.st]: Does this change anything?

**Review by guillep — COMMENTED** [review]
For the future we should check if the semantic analysis is a concern of the compiler or not, because many tools seem to rely on it.

File: src/OpalCompiler-Core/RBVariableNode.extension.st (hunk 1 above)
Comment (hogoww) [reply src/OpalCompiler-Core/RBVariableNode.extension.st:17]: On principle yes. But the compiler is not the only manipulating AST, and therefore let's call them "semantically analyses nodes".

File: src/GeneralRules/ReAssignmentOnBlockArgumentRule.class.st (hunk 2 above)
Comment (hogoww) [reply src/GeneralRules/ReAssignmentOnBlockArgumentRule.class.st]: Changes that if an ASTs is installed which has no binding, this wont throw an error.
I'm not sure if ASTs can be installed by someone else but the compiler, but I think we should still go with the right semantic method when we can..

**Comment by hogoww** [comment]
Agreed, that often comes in our discussions with Marcus, the later one being this morning

**Comment by dionisiydk** [comment]
I always had impression that changing the class of nodes during semantic analysis is a big smell.

Alternative would be to just delegate all semantics dependant questions to binding variable. Then we will not have such problems like here. There will be always same AST structure but we will inject specific bindings during semantics analysis. That would be a big change of course and probably it would have own drawbacks but it's good to think about alternatives.
From the other side AST-visitors can be kept compatible quite easily with extra delegation to bindings. And while most of ast users are visitors it could be not so hard change.

**Comment by Ducasse** [comment]
Yes guille was proposing the same: a little delegation.

**Comment by hogoww** [comment]
Removed conflict of updating a selector, which is not relevant anymore, as it the guilty class was removed.

this is now just a comment (cf first commit) about usage of #isArg.

**Comment by MarcusDenker** [comment]
I was looking at this finally.. sorry for not doing it before.

For #isArg, we should get rid of isArg before and replace it with #isArgument (no real change but less methods doing the same is good). They are both the result of the semantic analysis.

The new Variable hierarchy now is in sync with what we want: args are not temps and both are LocalVariables. 
That should make things much easier.

I will as a fist step deprecate it on the level of the Variable hierarchy and introduce isArgument there. Then we can fix the AST level in another step.

**Pull request closed without merging**